### PR TITLE
Add Concat to Sweeps

### PR DESCRIPTION
**Pull request by senecameeks**
A follow up PR will add Concat to cirq google sweep proto like in #6815

### CODE REVIEW

File: cirq-core/cirq/study/sweeps.py [hunk 1]
```diff
@@ -276,6 +276,64 @@ def _from_json_dict_(cls, factors, **kwargs):
...
+        if not isinstance(other, Concat):
+            return NotImplemented
+        return self.sweeps == other.sweeps
+
+    def __hash__(self):
+        return hash(tuple(self.sweeps))
+
+    @property
+    def keys(self) -> List['cirq.TParamKey']:
+        return self.sweeps[0].keys if self.sweeps else []
```
Comment (pavoljuhas) [inline cirq-core/cirq/study/sweeps.py:311]: Nit - self.sweeps are not empty per check in the `__init__`
```suggestion
        return self.sweeps[0].keys
```

Comment (senecameeks) [reply cirq-core/cirq/study/sweeps.py]: good catch!

File: cirq-core/cirq/study/sweeps_test.py [hunk 2]
```diff
@@ -373,3 +373,110 @@ def test_dict_to_zip_sweep():
...
+
+
+def test_concat_inconsistent_keys():
+    sweep1 = cirq.Linspace('a', 0, 1, 3)
+    sweep2 = cirq.Points('b', [2, 3])
+    try:
+        cirq.Concat(sweep1, sweep2)
+        assert False, "Expected ValueError for inconsistent keys"
+    except ValueError as e:
+        assert str(e) == "All sweeps must have the same descriptors."
```
Comment (pavoljuhas) [inline cirq-core/cirq/study/sweeps_test.py:431]:
```suggestion
    with pytest.raises(ValueError, match="All sweeps must have the same descriptors"):
        cirq.Concat(sweep1, sweep2)
```

File: cirq-core/cirq/study/sweeps_test.py [hunk 3]
```diff
@@ -373,3 +373,110 @@ def test_dict_to_zip_sweep():
...
+    sweep2 = cirq.Points('b', [2, 3])
+
+    with pytest.raises(ValueError, match="All sweeps must have the same descriptors."):
+        _ = cirq.Concat(sweep1, sweep2)
+
+
+def test_concat_empty_sweep_raises():
+    with pytest.raises(ValueError, match="Concat requires at least one sweep."):
+        _ = cirq.Concat()
+        _ = cirq.Concat([])
```
Comment (pavoljuhas) [inline cirq-core/cirq/study/sweeps_test.py:482]: the last line is never reached -
```suggestion
        _ = cirq.Concat()
```

**Review by pavoljuhas — CHANGES_REQUESTED** [review]
Please see inline comments and update tests to cover the changed lines here -
https://github.com/quantumlib/Cirq/actions/runs/12206286835/job/34056537224#step:7:1659

You may want to apply
```patch
diff --git a/cirq-core/cirq/study/sweeps_test.py b/cirq-core/cirq/study/sweeps_test.py
index 845da85c..b9097e23 100644
--- a/cirq-core/cirq/study/sweeps_test.py
+++ b/cirq-core/cirq/study/sweeps_test.py
@@ -246,6 +246,8 @@ def test_equality():
     et.make_equality_group(lambda: cirq.Linspace('b', 0, 10, 11))
     et.make_equality_group(lambda: cirq.Points('a', list(range(11))))
     et.make_equality_group(lambda: cirq.Points('b', list(range(11))))
+    et.make_equality_group(lambda: cirq.Concat(cirq.Linspace('a', 0, 10, 11)))
+    et.make_equality_group(lambda: cirq.Concat(cirq.Linspace('b', 0, 10, 11)))

     # Product and Zip sweeps can also be equated.
     et.make_equality_group(lambda: cirq.Linspace('a', 0, 5, 6) * cirq.Linspace('b', 10, 15, 6))
```
to cover the NotImplemented line here -
https://github.com/quantumlib/Cirq/blob/36d6e558caa5ec26af4df00b3f7e888d529b0a85/cirq-core/cirq/study/sweeps.py#L301-L303

Otherwise LGTM, thank you!

**Comment by senecameeks** [comment]
Thanks @pavoljuhas for the comments!

**Review by pavoljuhas — APPROVED** [review]
Looks, good, just please add a short test covering `Concat.keys`.
You can check coverage with `check/pytest-changed-files-and-incremental-coverage`
Thank you for adding this!

**Pull request merged**